src/hotspot/os/aix/os_perf_aix.cpp

```diff
@@ -22,519 +22,379 @@
-
-              18. priority %ld
-                     The standard nice value, plus fifteen.  The value is never negative in the kernel.
-
-              19. nice %ld
-                     The nice value ranges from 19 (nicest) to -19 (not nice to others).
-
-              20. 0 %ld  This value is hard coded to 0 as a placeholder for a removed field.
-
-              21. itrealvalue %ld
-                     The time in jiffies before the next SIGALRM is sent to the process due to an interval timer.
-
-              22. starttime %lu
-                     The time in jiffies the process started after system boot.
-
-              23. vsize %lu
-                     Virtual memory size in bytes.
-
-              24. rss %ld
-                     Resident Set Size: number of pages the process has in real memory, minus 3 for administrative purposes. This is just the pages which  count
-                     towards text, data, or stack space.  This does not include pages which have not been demand-loaded in, or which are swapped out.
-
-              25. rlim %lu
-                     Current limit in bytes on the rss of the process (usually 4294967295 on i386).
-
-              26. startcode %lu
-                     The address above which program text can run.
-
-              27. endcode %lu
-                     The address below which program text can run.
-
-              28. startstack %lu
-                     The address of the start of the stack.
-
-              29. kstkesp %lu
-                     The current value of esp (stack pointer), as found in the kernel stack page for the process.
-
-              30. kstkeip %lu
-                     The current EIP (instruction pointer).
-
-              31. signal %lu
-                     The bitmap of pending signals (usually 0).
-
-              32. blocked %lu
-                     The bitmap of blocked signals (usually 0, 2 for shells).
-
-              33. sigignore %lu
-                     The bitmap of ignored signals.
+#include <unistd.h>
 
-              34. sigcatch %lu
-                     The bitmap of catched signals.
+typedef struct {
+  u_longlong_t  user;
+  u_longlong_t  sys;
+  u_longlong_t  idle;
+  u_longlong_t  wait;
+} CPUPerfTicks;
 
-              35. wchan %lu
-                     This  is the "channel" in which the process is waiting.  It is the address of a system call, and can be looked up in a namelist if you need
-                     a textual name.  (If you have an up-to-date /etc/psdatabase, then try ps -l to see the WCHAN field in action.)
+typedef struct {
+  double utime;
+  double stime;
+} JVMTime;
 
-              36. nswap %lu
-                     Number of pages swapped - not maintained.
+enum {
+  UNDETECTED,
+  UNDETECTABLE,
+  LINUX26_NPTL,
+  BAREMETAL
+};
 
-              37. cnswap %lu
-                     Cumulative nswap for child processes.
+static bool initialize_libperfstat() {
+  static bool is_libperfstat_loaded = false;
```

Comment:
I don't think this is needed. libperfstat is initialized as part of os::init in os_aix.cpp.

https://github.com/openjdk/jdk/blob/d366d15d67a08833d93a5806edef8145cb7803e5/src/hotspot/os/aix/os_aix.cpp#L2999-L3008
https://github.com/openjdk/jdk/blob/d366d15d67a08833d93a5806edef8145cb7803e5/src/hotspot/os/aix/os_aix.cpp#L2329-L2333

Should be already active at this point. Its functions are stubbed out in case the library cannot be loaded, so you should be able to call them in here, without an additional init check. You just have to handle the errors on the individual API calls, as if the real APIs returned errors. But you do this already I think.

Long term, if you think libperfstat should be always there, we can simplify things and treat libperfstat load errors as hard VM errors in release builds too. We already assert in `os::Aix::initialize_libperfstat()`. (Note: "assert" fires in debug builds (ASSERT macro), and "guarantee" fires in all builds).

Reply:
Aha! I saw the definition of initialize_libperfstat(), but not that it was later called in os::init(). It did always feel strange to manually initialize libperfstat as I was doing. Thanks for pointing this out.